### PR TITLE
[Enhancement] optimize partition column statistics calculating

### PR DESCRIPTION
**Pull request by packy92**
## Why I'm doing:
optimize partition column statistics calculating
## What I'm doing:
- use the number of enumerate values  to adjust partition col ndv value for list partition table.
- use bucket info in histogram to estimate equal predicate filter ratio.

notation of modification in HistogramStatisticsTest:
- modifications of most cases are removing histogram in columnStatistics after join and some ndv values changed .
- modification of TPCH8 is mainly a join order changed
before :
```
|----22:EXCHANGE
|       distribution type: SHUFFLE
|       partition exprs: [46: C_CUSTKEY, INT, false]
|       cardinality: 3000000
|
11:EXCHANGE
distribution type: SHUFFLE
partition exprs: [37: O_CUSTKEY, INT, false]
cardinality: 1954166
```

after:
```
|----21:EXCHANGE
|       distribution type: SHUFFLE
|       partition exprs: [37: O_CUSTKEY, INT, false]
|       cardinality: 1323957
|
10:Project
|  output columns:
|  46 <-> [46: C_CUSTKEY, INT, false]
|  cardinality: 3000000
```
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Histogram.java [hunk 1]
```diff
@@ -57,4 +62,33 @@ public String getMcvString() {
...
+
+    public Optional<Long> getRowCountInBucket(ConstantOperator constantOperator, double distinctValuesCount) {
+        Optional<Double> value = StatisticUtils.convertStatisticsToDouble(constantOperator.getType(),
+                constantOperator.toString());
+        if (!value.isPresent()) {
+            return Optional.empty();
+        }
+
+        long rowCount;
+        for (int i = 0; i < buckets.size(); i++) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Histogram.java:74]:
```
public Optional<Long> getRowCountInBucket(ConstantOperator constantOperator, double distinctValuesCount) {
    Optional<Double> valueOpt = StatisticUtils.convertStatisticsToDouble(constantOperator.getType(), constantOperator.toString());
    if (!valueOpt.isPresent()) {
        return Optional.empty();
    }

    double value = valueOpt.get();

    int left = 0, right = buckets.size() - 1;
    while (left <= right) {
        int mid = (left + right) / 2;
        Bucket bucket = buckets.get(mid);
        
        if (bucket.getLower() <= value && value < bucket.getUpper()) {
            long rowCount = bucket.getCount() - bucket.getUpperRepeats();
            
            if (mid > 0) {
                rowCount -= buckets.get(mid - 1).getCount();
            }
            
            if (constantOperator.getType().isFixedPointType()) {
                rowCount = (long) Math.ceil(Math.max(1, rowCount / Math.max(1, (bucket.getUpper() - bucket.getLower()))));
            } else {
                rowCount = (long) Math.ceil(Math.max(1, rowCount / Math.max(1, distinctValuesCount / buckets.size())));
            }
            
            return Optional.of(rowCount);
        } else if (bucket.getUpper() == value) {
            return Optional.of(bucket.getUpperRepeats());
        } 
        
        if (value < bucket.getLower()) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    
    return Optional.empty();
}

```

The code from gpt, please consider.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/BinaryPredicateStatisticCalculator.java [hunk 2]
```diff
@@ -102,34 +101,48 @@ private static Statistics estimateColumnEqualToConstant(Optional<ColumnRefOperat
...
                     .build();
 
             double predicateFactor;
+            double rows;
+            Histogram hist = columnStatistic.getHistogram();
             Map<String, Long> histogramTopN = columnStatistic.getHistogram().getMCV();
-            // If there is a constant key in mcv, the ratio in mcv is directly used for filtering estimation.
-            // If it does not hit, filter out the key that appears in mcv, and then use the cardinality estimation
+
+            // If there is a constant key in mcv, we use the mvc count to estimate the row count.
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/BinaryPredicateStatisticCalculator.java:116]: mvc & mcv
should use MCV ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java [hunk 3]
```diff
@@ -701,64 +692,84 @@ public Void visitPhysicalJDBCScan(PhysicalJDBCScanOperator node, ExpressionConte
...
+                optimizerContext.getDumpInfo().addTableStatistics(olapTable,
+                        partitionCols.get(0).getName(),
+                        builder.getColumnStatistics(partitionCols.get(0)));
+            }
+            RangePartitionInfo rangePartitionInfo = (RangePartitionInfo) partitionInfo;
+            List<Map.Entry<Long, Range<PartitionKey>>> rangeList = null;
+            try {
+                rangeList = rangePartitionInfo.getSortedRangeMap(new HashSet<>(selectedPartitionId));
+            } catch (AnalysisException e) {
+                LOG.warn("get sorted range partition failed, msg : " + e.getMessage());
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:727]: that's really weird...

**Review by murphyatwork — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-3.2 branch-3.3

**Pull request merged**